File: bindings/typescript/src/context.ts
/**
 * PR-7 context-bundle views — a named, content-addressed collection a caller
 * attaches to a run (`invoke(handle, args, { context: [handle] })`) so a model
 * reasons over it. Kept in its own module so `types.ts` stays a thin aggregator
 * (the Rust core's module-per-concern discipline, GR3).
 *
 * SN-8: `bundleRef` is SERVER-DERIVED (blake3 over the manifest) — the client
 * names a handle, never an identity. The manifest lives in an off-journal
 * `bundles.db` sidecar (rebuildable-to-empty), scoped to the authoring party; a
 * not-found / not-owned bundle is UNIFORM (no cross-party existence oracle).
 */

import type { Id } from "./client.js";
import type {
  ContextBundle as PbContextBundle,
  ContextItem as PbContextItem,
  PutContextBundleResponse as PbPutContextBundleResponse,
} from "./gen/kortecx/v1/gateway_pb.js";
import { encode } from "./hexids.js";

/** One item to put in a context bundle: a label + a content-store ref. */
export interface ContextItemInput {
  /** Advisory label / context heading (display only). */
  name: string;
  /** A ref already in the content store (e.g. from {@link KxClientBase.putContent}). */
  contentRef: Id;
  /** Advisory mime (display / classify only). */
  mediaType?: string;
}

/** One item in a context bundle: an advisory label + a content-store ref. */
export class ContextBundleItem {
  constructor(
    readonly name: string,
    /** The 32-byte content-store ref, as 64 hex chars. */
    readonly contentRef: string,
    readonly mediaType: string,
  ) {}

  static fromProto(it: PbContextItem): ContextBundleItem {
    return new ContextBundleItem(it.name, encode(it.contentRef), it.mediaType);
  }

  toJSON() {
    return { name: this.name, content_ref: this.contentRef, media_type: this.mediaType };
  }
}

/** A context bundle's bound manifest (the governance / display view). */
export class ContextBundle {
  constructor(
    /** The server-derived manifest hash, as hex. */
    readonly bundleRef: string,
    readonly handle: string,
    readonly description: string,
    readonly items: ContextBundleItem[],
    readonly itemCount: number,
  ) {}

  static fromProto(b: PbContextBundle): ContextBundle {
    return new ContextBundle(
      encode(b.bundleRef),
      b.handle,
      b.description,
      b.items.map((it) => ContextBundleItem.fromProto(it)),
      b.itemCount,
    );
  }

  toJSON() {
    return {
      bundle_ref: this.bundleRef,
      handle: this.handle,
      description: this.description,
      item_count: this.itemCount,
      items: this.items.map((i) => i.toJSON()),
    };
  }
}

/** The outcome of a `PutContextBundle` upsert (server-derived ref + dedup flag). */
export class PutContextBundleResult {
  constructor(
    readonly bundleRef: string,
    readonly handle: string,
    readonly deduplicated: boolean,
  ) {}

  static fromProto(r: PbPutContextBundleResponse): PutContextBundleResult {
    return new PutContextBundleResult(encode(r.bundleRef), r.handle, r.deduplicated);
  }

  toJSON() {
    return { bundle_ref: this.bundleRef, handle: this.handle, deduplicated: this.deduplicated };
  }
}

File: bindings/python/src/kortecx/context.py
"""PR-7 context-bundle views — a named, content-addressed collection a caller
attaches to a run (``invoke(..., context=[handle])``) so a model reasons over it.

Kept in its own module so ``types.py`` stays a thin aggregator (the Rust core's
module-per-concern discipline, GR3). SN-8: ``bundle_ref`` is SERVER-DERIVED
(blake3 over the manifest) — the client names a handle, never an identity. The
manifest lives in an off-journal ``bundles.db`` sidecar (rebuildable-to-empty),
scoped to the authoring party; a not-found / not-owned bundle is UNIFORM (no
cross-party existence oracle).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List

from . import hexids
from .v1 import gateway_pb2 as _g


@dataclass(frozen=True)
class ContextBundleItem:
    """One item in a context bundle: an advisory label + a content-store ref."""

    name: str  # advisory label / context heading (display only)
    content_ref: str  # the 32-byte content-store ref, as 64 hex chars
    media_type: str  # advisory mime (display / classify only)

    @classmethod
    def from_proto(cls, it: "_g.ContextItem") -> "ContextBundleItem":
        return cls(
            name=it.name,
            content_ref=hexids.encode(it.content_ref),
            media_type=it.media_type,
        )


@dataclass(frozen=True)
class ContextBundle:
    """A context bundle's bound manifest (the governance / display view)."""

    bundle_ref: str  # server-derived manifest hash, as hex (16 bytes ⇒ 32 hex chars)
    handle: str  # the "namespace/collection/name" handle
    description: str  # advisory free-form
    items: List[ContextBundleItem]
    item_count: int

    @classmethod
    def from_proto(cls, b: "_g.ContextBundle") -> "ContextBundle":
        return cls(
            bundle_ref=hexids.encode(b.bundle_ref),
            handle=b.handle,
            description=b.description,
            items=[ContextBundleItem.from_proto(it) for it in b.items],
            item_count=b.item_count,
        )


@dataclass(frozen=True)
class PutContextBundleResult:
    """The outcome of a ``PutContextBundle`` upsert (server-derived ref + dedup)."""

    bundle_ref: str  # server-derived manifest hash, as hex
    handle: str  # echoed canonical handle
    deduplicated: bool  # an identical manifest was already bound to this handle

    @classmethod
    def from_proto(cls, r: "_g.PutContextBundleResponse") -> "PutContextBundleResult":
        return cls(
            bundle_ref=hexids.encode(r.bundle_ref),
            handle=r.handle,
            deduplicated=r.deduplicated,
        )
